### PR TITLE
feat: remove expired form submissions

### PR DESCRIPTION
**Pull request by laurensmertens**
|Q              |A  |
|---------------|---|
|Bug fix?       |No|
|New feature?   |Yes|	
|BC breaks?     |No|	
|Deprecations?  |No|	
|Fixed tickets  |No|

Added a method in the FormSubmissionService to remove expired submissions and a Command to use it.

### CODE REVIEW

File: Entity/FormSubmission.php [hunk 1]
```diff
@@ -187,7 +187,7 @@ public function getLabel(): string
         return $this->label;
     }
 
-    public function getDeadlineDate(): string
+    public function getDeadlineDate(): ?string
     {
```
Comment (theus77) [inline Entity/FormSubmission.php:191]: Why does a date is in a string type?

File: Repository/FormSubmissionRepository.php [hunk 2]
```diff
@@ -40,4 +40,10 @@ public function save(FormSubmission $formSubmission): void
         $this->_em->persist($formSubmission);
         $this->_em->flush();
     }
+
+    public function remove(FormSubmission $formSubmission): void
+    {
+        $this->_em->remove($formSubmission);
+        $this->_em->flush();
```
Comment (theus77) [inline Repository/FormSubmissionRepository.php:47]: I'll check but why will you flush all pending updates here?

File: Service/Form/Submission/FormSubmissionService.php [hunk 3]
```diff
@@ -91,4 +91,24 @@ public function submit(FormSubmissionRequest $submitRequest): array
 
         return ['submission_id' => $formSubmission->getId()];
     }
+
+    public function removeExpiredSubmissions(): int
+    {
+        $removedCount = 0;
+
+        foreach ($this->repository->findAll() as $submission) {
+            $deadline = $submission->getDeadlineDate();
```
Comment (theus77) [inline Service/Form/Submission/FormSubmissionService.php:100]: The repository should implement a function like getAllOutdatedSubmission?

File: Service/Form/Submission/FormSubmissionService.php [hunk 4]
```diff
@@ -91,4 +91,24 @@ public function submit(FormSubmissionRequest $submitRequest): array
...
+            if (!\is_string($deadline)) {
+                continue;
+            }
+
+            $expireTimestamp = \strtotime($deadline);
```
Comment (theus77) [inline Service/Form/Submission/FormSubmissionService.php:105]: In the get getDeadlineDate function?

File: Entity/FormSubmission.php (hunk 1 above)
Comment (dameert) [reply Entity/FormSubmission.php:191]: Nice catch, I overlooked this earlier!

File: Service/Form/Submission/FormSubmissionService.php (hunk 3 above)
Comment (Davidmattei) [reply Service/Form/Submission/FormSubmissionService.php:100]: Agree

Comment (dameert) [reply Service/Form/Submission/FormSubmissionService.php:100]: indeed, this function should call respository->findAllOutdatedSubmissions($expirationDate).

And then delete everything that is returned.

File: Service/Form/Submission/FormSubmissionService.php (hunk 4 above)
Comment (dameert) [reply Service/Form/Submission/FormSubmissionService.php:105]: Deadline should not be a string but DB  in remark, so this conversion should not be needed.

File: Command/RemoveExpiredSubmissionsCommand.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace EMS\CoreBundle\Command;
+
+use Elasticsearch\Client;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use EMS\CoreBundle\Service\Form\Submission\FormSubmissionService;
+
+class RemoveExpiredSubmissionsCommand extends EmsCommand
```
Comment (Davidmattei) [inline Command/RemoveExpiredSubmissionsCommand.php:11]: Don't think you  should extend from EmsCommand.
EmsCommand will wait for elasticsearch to be  green, but for this command we don't need elasticsearch. 

Just extend from Symfony\Component\Console\Command\Command and implement EMS\CommonBundle\Command\CommandInterface

**Review by dameert — CHANGES_REQUESTED** [review]
see comments of mdk

**Review by Davidmattei — CHANGES_REQUESTED** [review]

File: Service/Form/Submission/FormSubmissionService.php (hunk 3 above)
Comment (laurensmertens) [reply Service/Form/Submission/FormSubmissionService.php:100]: @dameert Do we need the expirationDate argument? We just need to find those who are in the past right?

File: Repository/FormSubmissionRepository.php (hunk 2 above)
Comment (laurensmertens) [reply Repository/FormSubmissionRepository.php:47]: OK I'll create a separate flush function and only call it once.

File: Entity/FormSubmission.php (hunk 1 above)
Comment (laurensmertens) [reply Entity/FormSubmission.php:191]: OK, this was to allow any date format, but indeed not a good idea.

**Review by Davidmattei — APPROVED** [review]

File: Repository/FormSubmissionRepository.php [hunk 6]
```diff
@@ -7,6 +7,7 @@
 use Doctrine\Bundle\DoctrineBundle\Registry;
 use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
 use EMS\CoreBundle\Entity\FormSubmission;
+use phpDocumentor\Reflection\Types\Void_;
```
Comment (Davidmattei) [inline Repository/FormSubmissionRepository.php:10]: don't like this

**Review by Davidmattei — CHANGES_REQUESTED** [review]

File: Repository/FormSubmissionRepository.php (hunk 6 above)
Comment (laurensmertens) [reply Repository/FormSubmissionRepository.php:10]: I removed it, it was added by PHPStorm by accident.

File: Resources/DoctrineMigrations/pdo_mysql/Version20200921075641.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+        $this->addSql('ALTER TABLE form_submission DROP COLUMN deadline_date ;');
+        $this->addSql('ALTER TABLE form_submission ADD deadline_date DATE DEFAULT NULL');
+    }
+
+    public function down(Schema $schema) : void
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+    }
```
Comment (theus77) [inline Resources/DoctrineMigrations/pdo_mysql/Version20200921075641.php:23]: Down queries are missing

File: Resources/DoctrineMigrations/pdo_pgsql/Version20200921075641.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    public function up(Schema $schema) : void
+    {
+        $this->addSql('ALTER TABLE form_submission DROP COLUMN deadline_date ;');
+        $this->addSql('ALTER TABLE form_submission ADD deadline_date DATE DEFAULT NULL');
+    }
+
+    public function down(Schema $schema) : void
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'postgresql', 'Migration can only be executed safely on \'postgresql\'.');
+    }
```
Comment (theus77) [inline Resources/DoctrineMigrations/pdo_pgsql/Version20200921075641.php:22]: Down queries are missing

File: Resources/DoctrineMigrations/pdo_sqlite/Version20200921080015.php [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'sqlite', 'Migration can only be executed safely on \'sqlite\'.');
+        $this->addSql('ALTER TABLE form_submission DROP COLUMN deadline_date ;');
+        $this->addSql('ALTER TABLE form_submission ADD deadline_date DATE DEFAULT NULL');
+    }
+
+    public function down(Schema $schema) : void
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'sqlite', 'Migration can only be executed safely on \'sqlite\'.');
+    }
```
Comment (theus77) [inline Resources/DoctrineMigrations/pdo_sqlite/Version20200921080015.php:23]: Down queries are missing

File: Service/Form/Submission/FormSubmissionRequest.php [hunk 10]
```diff
@@ -22,7 +22,7 @@ final class FormSubmissionRequest
     private $files;
     /** @var string */
     private $label;
-    /** @var string */
+    /** @var \DateTime|false */
```
Comment (theus77) [inline Service/Form/Submission/FormSubmissionRequest.php:25]: false or null?

File: Service/Form/Submission/FormSubmissionRequest.php [hunk 11]
```diff
@@ -80,9 +80,9 @@ public function getLabel(): string
         return $this->label;
     }
 
-    public function getDeadlineDate(): string
+    public function getDeadlineDate(): ?\DateTime
     {
-        return $this->deadlineDate;
+        return $this->deadlineDate !== false ? $this->deadlineDate : null;
```
Comment (theus77) [inline Service/Form/Submission/FormSubmissionRequest.php:85]: It is never assigne to false (as far as I have reviewed your code)

File: Service/Form/Submission/FormSubmissionRequest.php (hunk 10 above)
Comment (laurensmertens) [reply Service/Form/Submission/FormSubmissionRequest.php:25]: I had null at first but Travis wouldn't allow it.

File: Service/Form/Submission/FormSubmissionRequest.php (hunk 11 above)
Comment (laurensmertens) [reply Service/Form/Submission/FormSubmissionRequest.php:85]: Same here, Travis gives an error like "deadline date cannot be DateTime|false|null".

File: Repository/FormSubmissionRepository.php [hunk 12]
```diff
@@ -35,9 +35,39 @@ public function findAllUnprocessed(): array
         return $qb->getQuery()->getArrayResult();
     }
 
+    public function removeAllOutdatedSubmission(): int
+    {
+        $outdatedSubmissions = $this->createQueryBuilder('fs')
+            ->andWhere('fs.expireDate < :olderThan')
```
Comment (dameert) [inline Repository/FormSubmissionRepository.php:41]:
```suggestion
            ->andWhere('fs.expireDate <= :today')
```

**Review by dameert — CHANGES_REQUESTED** [review]

File: Service/Form/Submission/FormSubmissionRequest.php [hunk 13]
```diff
@@ -41,7 +41,7 @@ public function __construct(Request $request)
         $this->data = $submit['data'];
         $this->files = $submit['files'];
         $this->label = $submit['label'] ?? '';
-        $this->deadlineDate = $submit['deadline_date'] ?? '';
+        $this->deadlineDate = \DateTime::createFromFormat('m/d/Y', $submit['deadline_date']) ?? null;
     }
```
[the remark this answers is not in the document]
Comment (Davidmattei) [reply Service/Form/Submission/FormSubmissionRequest.php:46]: \DateTime::ATOM

**Pull request closed without merging**